File: testing/smoothing/smoothing_calc.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

// Function to calculate the smooth transition (based on the original C expression)
double smooth_transition(double s1, double s2, int ramp_step, int T) {
    return s1 + ((s2 - s1) * 0.5 * (1.0 - cos(ramp_step * M_PI / T)));
}

int main(int argc, char *argv[]) {
    // Check that the correct number of command-line arguments is provided
    if (argc != 4) {
        printf("Usage: %s <s1> <s2> <T>\n", argv[0]);
        return 1;
    }

    // Parse command-line arguments
    double s1 = atof(argv[1]);  // Starting value
    double s2 = atof(argv[2]);  // Ending value
    int T = atoi(argv[3]);     // Number of steps

    printf("S1=%f  S2=%f  T=%d\n",s1,s2,T);
    // Open the output file for writing
    FILE *file = fopen("smooth_output.txt", "w");
    if (file == NULL) {
        perror("Error opening file");
        return 1;
    }

    // Write the header to the file
    fprintf(file, "Ramp Step, Transition Value\n");

    // Calculate the smooth transition for each ramp step and write to the file
    for (int ramp_step = 0; ramp_step <= T; ramp_step++) {
        double transition = smooth_transition(s1, s2, ramp_step, T);
        fprintf(file, "%d, %.6f\n", ramp_step, transition);
    }

    // Close the file
    fclose(file);

    printf("Output written to smooth_output.txt\n");

    return 0;
}

File: testing/smoothing/smoothing_calc.py
import numpy as np
import matplotlib.pyplot as plt
import argparse

# Function to calculate the smooth transition (based on the original C expression)
def smooth_transition(s1, s2, ramp_step, T):
    return s1 + ((s2 - s1) * 0.5 * (1.0 - np.cos(ramp_step * np.pi / T)))

# Main function
def main():
    # Argument parser for command-line parameters
    parser = argparse.ArgumentParser(description="Plot smooth transition between s1 and s2.")
    parser.add_argument('s1', type=float, help="Starting value (s1)")
    parser.add_argument('s2', type=float, help="Ending value (s2)")
    parser.add_argument('T', type=int, help="T value (total number of steps)")
    
    # Parse the arguments
    args = parser.parse_args()
    
    s1 = args.s1
    s2 = args.s2
    T = args.T
    
    # Generate values for ramp_step from 0 to T
    ramp_steps = np.arange(0, T+1)
    
    # Calculate the smooth transition for each ramp step
    transitions = [smooth_transition(s1, s2, step, T) for step in ramp_steps]
    
    # Plot the results
    plt.plot(ramp_steps, transitions, label=f'Smooth transition from {s1} to {s2}')
    plt.xlabel('Ramp Step')
    plt.ylabel('Transition Value')
    plt.title(f'Smooth Transition: {s1} to {s2}')
    plt.grid(True)
    plt.legend()
    plt.show()

# Execute the script
if __name__ == "__main__":
    main()

